### PR TITLE
Fixes race condition when debugging Lua files

### LINKED ISSUE #14823: LuaIDE Breakpoint Step Marker randomly goes out of sync when "Lua Locals" are enabled.
**Describe the bug**
When debugging Lua scripts with LuaIDE.exe, and "Lua Locals" are enabled, the  debug step marker stops moving, even though you may still "Step Over" lines of code. When LuaIDE.exe enters in this bad state, you'll have to remove the breakpoints, run "Continue", and detach from the Editor. And kill the LuaIDE.exe to start anew.

**Assets required**
[LuaLocalsTest.zip](https://github.com/o3de/o3de/files/10834375/LuaLocalsTest.zip)
Unzip the attached level into `<Game>/Levels/` folder.

**Steps to reproduce**
1. In the Editor.exe Open the attached `LuaLocalsTest` Level, let's say in the AutomatedTesting/ game project.
2. In the `Entity Outliner`, select the **EntityMover** entity, In the **Lua Script** component click the `Open In Lua Editor` button.
3. Wait until LuaIDE.exe loads `entity_mover.lua`.
4. In LuaIDE.exe attach to Editor.exe.
5. In LuaIDE.exe click the `Lua Locals` button which should open the `Lua Locals` panel.
6. Place a breakpoint in line 36 of `entity_mover.lua` (just below `function entity_mover:OnTick(deltaTime, timePoint)`).
7. In Editor.exe enter game mode (Ctrl + G).
8. The LuaIDE should regain focus and the red Breakpoint should be in the same position as the "Breakpoint Step Marker" (green triangle).
9. Click "Step Over" until the   "Breakpoint Step Marker" (green triangle) stops moving and LuaIDE.exe becomes visually unresponsive. Sometimes you may end up stepping 5 times over the "OnTick" function, but most of the time within a few "Step Over" events the bug occurs.

**Expected behavior**
You should be able to "Step Over" as needed and the "Breakpoint Step Marker" (green triangle) should move to the next line as you "Step Over".

**Actual behavior**
Eventually  the "Breakpoint Step Marker" (green triangle) stops moving to the next line as you click "Step Over"

**Screenshots/Video**
In this picture, the LuaIDE is already stuck. The  "Breakpoint Step Marker" (green triangle)  doesn't move anymore and and the "Lua Locals" panel stops updating.
image

**Found in Branch**
commit https://github.com/o3de/o3de/commit/76074304573950182e2d48283ff4b01b798bab03 (upstream/development, development)
Author: lumberyard-employee-dm [56135373+lumberyard-employee-dm@users.noreply.github.com](mailto:56135373+lumberyard-employee-dm@users.noreply.github.com)
Date: Fri Feb 24 17:19:34 2023 -0600

**Desktop/Device (please complete the following information):**
    Device: PC
    OS: Windows
    Version 11
    CPU AMD Ryzen Threadripper 3970X 32-Core Processor 3.70 GHz
    GPU NVidia RTX 4090 (24GB VRAM)
    Memory 128GB
    Storage: WD_BLACK 2TB SN850 NVMe M2 SDD PCIe 4.0

**Additional context**
It is very important to have the `Lua Locals` panel open for this bug to reproduce.

### PR DESCRIPTION
**Pull request by galibzon**
Closes #14823

## What does this PR do?

The TcpConnection class has a m_sendRingBuffer
that is NOT thread safe.

NetworkSystemComponent uses the Main thread
to send messages, and hence use the m_sendRingBuffer.

On the other hand, before this fix, the RemoteToolsSystemComponent.cpp was sending packets with the help of a second thread controlled by the class RemoteToolsOutboxThread.cpp. This second thread is totally unnecessary and, even worst, was causing race conditions. This race condition was corrupting the messages being sent from Editor.exe to LuaIDE.exe.

## How was this PR tested?

Did long sessions of Lua files testing. For small and large Lua files. No problems at all with this fix.
Without this fix the LuaIDE is mostly unusable.

## Other Notes.
In particular the Race Condition exacerbates when this PR was merged: https://github.com/o3de/o3de/pull/10132
That PR was necessary, but it forces the TcpConnection to flush pending messages in the Main thread, while at the same time the RemoteToolsSystemComponent was also sending messages using the RemoteToolsOutboxThread.

Here is an example of the Race Condition:

**Main Thread**  
```
 	EditorLib.dll!AzNetworking::TcpRingBufferImpl::AdvanceReadBuffer(unsigned int numBytes) Line 54	C++
 	EditorLib.dll!AzNetworking::TcpRingBuffer<1048576>::AdvanceReadBuffer(unsigned int numBytes) Line 48	C++
 	EditorLib.dll!AzNetworking::TcpConnection::UpdateSend() Line 112	C++
 	EditorLib.dll!AzNetworking::TcpNetworkInterface::HandleConnectionSend(AzNetworking::SocketFd socketFd) Line 219	C++
 	EditorLib.dll!AzNetworking::TcpNetworkInterface::Update::__l2::<lambda_2>::operator()(AzNetworking::SocketFd socketFd) Line 115	C++
 	EditorLib.dll!AZStd::Internal::INVOKE<`AzNetworking::TcpNetworkInterface::Update'::`2'::<lambda_2> &,enum AzNetworking::SocketFd>(AzNetworking::TcpNetworkInterface::Update::__l2::<lambda_2> & f, AzNetworking::SocketFd && <args_0>) Line 209	C++
 	EditorLib.dll!AZStd::invoke<`AzNetworking::TcpNetworkInterface::Update'::`2'::<lambda_2> &,enum AzNetworking::SocketFd>(AzNetworking::TcpNetworkInterface::Update::__l2::<lambda_2> & f, AzNetworking::SocketFd && <args_0>) Line 55	C++
 	EditorLib.dll!AZStd::Internal::function_util::invoke_void_return_wrapper<void>::call<`AzNetworking::TcpNetworkInterface::Update'::`2'::<lambda_2> &,enum AzNetworking::SocketFd>(AzNetworking::TcpNetworkInterface::Update::__l2::<lambda_2> & <args_0>, AzNetworking::SocketFd && <args_1>) Line 42	C++
 	EditorLib.dll!AZStd::Internal::function_util::get_invoker<void __cdecl(enum AzNetworking::SocketFd),AZStd::Internal::function_util::function_obj_tag,AZStd::allocator>::call<`AzNetworking::TcpNetworkInterface::Update'::`2'::<lambda_2>>(AZStd::Internal::function_util::function_buffer & function_obj_ptr, AzNetworking::SocketFd && <args_0>) Line 173	C++
 	EditorLib.dll!AZStd::function_intermediate<void,enum AzNetworking::SocketFd>::operator()(AzNetworking::SocketFd && <args_0>) Line 605	C++
 	EditorLib.dll!AZStd::function<void __cdecl(enum AzNetworking::SocketFd)>::operator()(AzNetworking::SocketFd <args_0>) Line 685	C++
 	EditorLib.dll!AzNetworking::TcpSocketManager::ProcessEvents(AZ::TimeMs maxBlockMs, const AZStd::function<void __cdecl(enum AzNetworking::SocketFd)> & readCallback, const AZStd::function<void __cdecl(enum AzNetworking::SocketFd)> & writeCallback) Line 75	C++
 	EditorLib.dll!AzNetworking::TcpNetworkInterface::Update() Line 116	C++
 	EditorLib.dll!AzNetworking::NetworkingSystemComponent::OnSystemTick() Line 82	C++
 	EditorLib.dll!AzNetworking::NetworkingSystemComponent::ForceUpdate() Line 175	C++
>	EditorLib.dll!AzFramework::ScriptDebugAgent::BreakpointCallback(AZ::ScriptContextDebug * debugContext, const AZ::ScriptContextDebug::Breakpoint * breakpoint) Line 448	C++

```

**RemoteTools::RemoteToolsOutboxThread**  
```
>	EditorLib.dll!AzNetworking::TcpConnection::SendPacketInternal(AzNetworking::PacketType packetType, AzNetworking::ByteBuffer<16384> & payloadBuffer, AZ::TimeMs currentTimeMs) Line 347	C++
 	EditorLib.dll!AzNetworking::TcpConnection::SendReliablePacket(const AzNetworking::IPacket & packet) Line 207	C++
 	EditorLib.dll!AzNetworking::TcpNetworkInterface::SendReliablePacket(AzNetworking::ConnectionId connectionId, const AzNetworking::IPacket & packet) Line 133	C++
 	RemoteTools.dll!RemoteTools::RemoteToolsOutboxThread::OnUpdate(AZ::TimeMs updateRateMs) Line 85	C++
 	RemoteTools.dll!AzNetworking::TimedThread::Start::__l2::<lambda_1>::operator()() Line 33	C++
 	RemoteTools.dll!AZStd::Internal::INVOKE<`AzNetworking::TimedThread::Start'::`2'::<lambda_1>>(AzNetworking::TimedThread::Start::__l2::<lambda_1> && f) Line 209	C++
 	RemoteTools.dll!AZStd::invoke<`AzNetworking::TimedThread::Start'::`2'::<lambda_1>>(AzNetworking::TimedThread::Start::__l2::<lambda_1> && f) Line 55	C++
 	RemoteTools.dll!AZStd::Internal::apply_impl<`AzNetworking::TimedThread::Start'::`2'::<lambda_1>,std::tuple<>>(AzNetworking::TimedThread::Start::__l2::<lambda_1> && f, std::tuple<> && tupleObj, std::integer_sequence<unsigned __int64> __formal) Line 282	C++
 	RemoteTools.dll!AZStd::apply<`AzNetworking::TimedThread::Start'::`2'::<lambda_1>,std::tuple<>>(AzNetworking::TimedThread::Start::__l2::<lambda_1> && f, std::tuple<> && tupleObj) Line 290	C++
 	RemoteTools.dll!AZStd::thread::{ctor}::__l2::<lambda_1>::operator()() Line 46	C++
 	RemoteTools.dll!AZStd::Internal::thread_info_impl<`AZStd::thread::thread<`AzNetworking::TimedThread::Start'::`2'::<lambda_1>>'::`2'::<lambda_1>>::execute() Line 193	C++
 	RemoteTools.dll!AZStd::Internal::thread_run_function(void * param) Line 38	C++


```

### CODE REVIEW

**Review by lumberyard-employee-dm — APPROVED** [review]

File: Gems/RemoteTools/Code/Source/RemoteToolsSystemComponent.cpp [hunk 1]
```diff
@@ -363,17 +361,26 @@ namespace RemoteTools
...
+        size_t outSize = totalSize;
+        while (outSize > 0 && networkInterface != nullptr)
         {
-            m_outboxThread->Join();
-            m_outboxThread->Start();
+            // Fragment the message into RemoteToolsMessageBuffer packet sized chunks and send
+            size_t bufferSize = AZStd::min(outSize, aznumeric_cast<size_t>(RemoteToolsBufferSize));
+            RemoteToolsPackets::RemoteToolsMessage tmPacket;
+            tmPacket.SetPersistentId(target.GetPersistentId());
+            tmPacket.SetSize(aznumeric_cast<uint32_t>(totalSize));
```
Comment (moraaar) [inline Gems/RemoteTools/Code/Source/RemoteToolsSystemComponent.cpp:375]: Just checking, the size of the `tmPacket` in every loop is always the full size `totalSize` and not the size of the current message buffer `bufferSize`?

**Review by moraaar — APPROVED** [review]
Apart from the question, below it looks good to me. Great job finding and fixing the race condition!

File: Gems/RemoteTools/Code/Source/RemoteToolsSystemComponent.cpp (hunk 1 above)
Comment (galibzon) [reply Gems/RemoteTools/Code/Source/RemoteToolsSystemComponent.cpp:375]: Even though the message may get broken into several pieces of maximum `16,000` bytes, the tmPacket.SetSize() must be the TotalSize. In other words, all of the submessages carry the TotalSize with them, in addition the message owns a subpacket encoded as RemoteToolsMessageBuffer, which has the subpacket size, in this case `bufferSize`.

Long story short, ` tmPacket.SetSize(aznumeric_cast<uint32_t>(totalSize));` is correct.

Comment (moraaar) [reply Gems/RemoteTools/Code/Source/RemoteToolsSystemComponent.cpp:375]: Thanks for the clarification.

File: Gems/RemoteTools/Code/Source/RemoteToolsSystemComponent.cpp [hunk 2]
```diff
@@ -363,17 +361,26 @@ namespace RemoteTools
...
+            RemoteToolsMessageBuffer encodingBuffer;
+            encodingBuffer.CopyValues(outBuffer + (totalSize - outSize), bufferSize);
+            tmPacket.SetMessageBuffer(encodingBuffer);
+            outSize -= bufferSize;
+
+            networkInterface->SendReliablePacket(connectionId, tmPacket);
```
Comment (jackalbe) [inline Gems/RemoteTools/Code/Source/RemoteToolsSystemComponent.cpp:381]: Should this continue to loop even if the SendReliablePacket() fails?

Comment (galibzon) [reply Gems/RemoteTools/Code/Source/RemoteToolsSystemComponent.cpp:381]: Good point, will add an error check for that.

**Review by lsemp3d — APPROVED** [review]

**Pull request merged**